image/image.go

```diff
@@ -50,17 +50,31 @@ var validRefMediaTypes = []string{
 }
 
 func validate(w walker, refs []string, out *log.Logger) error {
-	for _, r := range refs {
-		ref, err := findDescriptor(w, r)
-		if err != nil {
-			return err
+	ds, err := findDescriptors(w)
+	if err != nil {
+		return err
+	}
+	// TODO(runcom): this is O(N) but it could be O(1) if refs have been
+	// map[string]bool coming from the cli package.
+	// However, I don't think refs could be large enough to impact performance.
+	inRefs := func(ref string) bool {
+		for _, r := range refs {
+			if r == ref {
+				return true
+			}
 		}
-
-		if err = ref.validate(w, validRefMediaTypes); err != nil {
+		return false
+	}
+	hasRefs := len(refs) != 0
+	for ref, d := range ds {
+		if hasRefs && !inRefs(ref) {
```

Comment:
On Thu, Sep 08, 2016 at 08:40:01AM -0700, Antonio Murdaca wrote:
> >  		}
> > -
> > -		if err = ref.validate(w, validRefMediaTypes); err != nil {
> > +		return false
> > +	}
> > +	hasRefs := len(refs) != 0
> > +	for ref, d := range ds {
> > +		if hasRefs && !inRefs(ref) {
> 
> it's equivalent right? you still do an O(N) iterating over refs :(

Iterating over refs is O(len(refs)).  Iterating over ds is O(len(ds)).
‘refs’ is going to be shorter than or the same length as ‘ds’ unless
someone called validate --refs 1.0 --refs 1.0 --refs 1.0 (or some
such).  I don't expect that to happen frequently enough to be worth
deduplicating refs (although iterating over ‘ds’ is one way to
deduplicate ‘refs’).

My main reason for wanting to iterate over refs is that it runs the
validation in the same order that the caller asked for.